### DISCUSSION TITLE
Why MinIO uses 1.6 billion inodes a day?

### DISCUSSION THREAD

**Discussion by adababys** [minio/minio#19630] [Q&A] [answered]
![inode_info](https://github.com/minio/minio/assets/8872412/db3da20a-4b80-45d9-8189-329f9cca64b0)

Currently, the MinIO cluster has 3 nodes with 8 disks each, and the cluster has stored 4.5TB of data, using 1.6 billion inodes. I want to check where these 1.6 billion inodes are being used and whether this is normal.

Stable traffic with only tens of millions of object writes per day, with a lifecycle set to 1 day.

Who can help check the usage of inodes? If you need any other information, please feel free to contact me. Thank you.

**Comment by adababys** [comment]
The cluster has a total of 40 buckets, 6.8 million files, with an average file size of 150K

**Comment by adababys** [comment]
![delete-file-but-not-inode](https://github.com/minio/minio/assets/8872412/3da1876f-bd6d-4a0b-b272-192aaa84bf1d)
The file was deleted after reaching its lifecycle expiration, but the inode still remains. I believe this might be a bug that needs to be reported.

**Comment by klauspost** [comment, marked as answer]
MinIO does not manage inodes; your filesystem does. For performance related questions look into our [subscriptions](https://min.io/pricing).

**Comment by adababys** [comment]
When I use the bpftrace tool to debug the kernel, I find that MinIO keeps opening non-existent files, which prevents inodes from being released. What could be the reason, and how should I investigate this?
![bpftrace-minio-open-file](https://github.com/minio/minio/assets/8872412/602f228d-4c6b-40d2-bfa5-7c1f1b6f0ecb)

**Comment by ghost** [comment]
I am seeing a similar issue. I have been digging through a number of issues on the minio repo and reading about how minio uses temporary files during and after uploads.

I realize that the filesystem manages inodes, but I think what's confusing some of us here is how **minio** uses **the filesystem** (which creates/holds inodes). For example, in the above bpftrace results it shows inodes not being released _because_ minio is trying to access certain files that are long gone. If that's accurate (and that really is what's happening), then to me that's not good behaviour. It won't be sustainable in production - how many billions of inodes should this person have in order to let minio run? Eventually it would run out from being swamped by "phantom file accessses", right?

I'll explain about my situation to show what I mean:

- Very simple use case: we run minio on a small disk with a relatively low number of inodes in the filesystem.
- 4 million inodes available in the filesystem, and we know that we will be storing a static set of 1.1 million small files. Files are each <2kB in size, so our dataset is on the order of a couple gigabytes (but _many_ small files).
- The files are so small that we would run out of inodes long before we run out of blocks in the filesystem, if we "scaled up"
- We notice that when we upload our set of files to minio, the number of inodes used is always **double** the number of files. We ingest 1.1M files and have 2.2M inodes consumed afterward (`df -i` output).
- After more than 24 hours, we do not see this count coming down, but I am not able to see using tools like `find`, `stat`, `du` etc where the extra inodes are coming from (some kind of symlinks/dotfiles created by minio, perhaps?)

My situation is similar to this thread: @adababys has a system that has 6.8M active files, with on the order of 10M object writes per day, with a 1 day lifecycle ... but there are 1.6B inodes held. "Back of the napkin" math tells us that something is not right with that ("inodes taken" >>>>> "expected number of writes+active files within 1 day window" by about a factor of 100?).

If I stop my minio process and fully clear the path it's using for its data, the inode count comes back down to what I expect. Obviously this is not an option for me - I want to keep minio operating, but just not have "double the inodes" consumed compared to what I expect.

**Saying this as respectfully as I can:** I don't consider this a performance problem. I don't think it's the fault of the filesystem that we have so many inodes being taken. I would like to know more about what's going on, so I can:

- configure something about minio to control this more closely
OR
- be able to say "the total inodes consumed will never rise above X as long as we use the system the way we are planning to"

Alternatively, if it is the fault of the filesystem: what is my completely stock and up-to-date ext4 doing that's incorrect?

Does that make sense? I will keep doing some more reading about this to see what else I can find, but so far I haven't found answers in the following threads:

- https://github.com/minio/minio/issues/19632
- https://github.com/minio/minio/issues/5222
- https://github.com/minio/minio/issues/14713
- https://github.com/minio/minio/issues/14715
- https://github.com/minio/minio/issues/8129

These are all from different time periods, and they concern different minio versions and different client setups.. but they seem related to what this discussion is about: we all want to understand and control inode consumption on our minio instances.

Hopefully you see what I mean! Any advice for debugging/understanding/configuring this would be appreciated.
